{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GA Implementation\n",
    "\n",
    "https://medium.com/analytics-vidhya/a-genetic-algorithm-for-optimizing-neural-network-parameters-d8187d5114ed\n",
    "\n",
    "Here are quick steps for how the genetic algorithm works:\n",
    "\n",
    "-Initial Population– Initialize the population randomly based on the data.\n",
    "\n",
    "-Fitness function– Find the fitness value of the each of the chromosomes(a chromosome is a set of parameters which define a proposed solution to the problem that the genetic algorithm is trying to solve)\n",
    "\n",
    "-Selection– Select the best fitted chromosomes as parents to pass the genes for the next generation and create a new population\n",
    "\n",
    "Cross-over– Create new set of chromosome by combining the parents and add them to new population set\n",
    "\n",
    "Mutation– Perfrom mutation which alters one or more gene values in a chromosome in the new population set generated. Mutation helps in getting more diverse oppourtinity.Obtained population will be used in the next generation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import libraries\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from random import choice, uniform, random, randint\n",
    "from operator import itemgetter, attrgetter\n",
    "\n",
    "# Second, import deep learning library\n",
    "import keras\n",
    "from keras import backend as K\n",
    "from keras.models import Sequential\n",
    "from keras.layers import Conv2D,MaxPooling2D,Dropout,Activation,Flatten,Dense\n",
    "from keras.optimizers import Adam\n",
    "from keras.preprocessing.image import ImageDataGenerator\n",
    "import os\n",
    "from tensorflow.keras import backend\n",
    "from tensorflow.keras.applications import VGG16, MobileNet\n",
    "#Import packages used here:\n",
    "# for initial data exploration:\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "from IPython.display import Image, display\n",
    "import random\n",
    "import math\n",
    "\n",
    "#For modeling and model viewing. \n",
    "import tensorflow as tf\n",
    "from keras.preprocessing.image import load_img,img_to_array,ImageDataGenerator\n",
    "from tensorflow.keras.models import Model, Sequential\n",
    "from tensorflow.keras.utils import plot_model \n",
    "from tensorflow.keras.layers import Input, Dense, Dropout, Flatten, Activation,Concatenate\n",
    "from tensorflow.keras.layers import Conv2D, MaxPooling2D, AveragePooling2D\n",
    "from tensorflow.keras.layers import BatchNormalization\n",
    "from tensorflow.keras.utils import to_categorical #Image generator used for transformation to categorical\n",
    "from tensorflow.keras.optimizers import Adam, SGD\n",
    "from tensorflow.keras.callbacks import EarlyStopping, LearningRateScheduler, ReduceLROnPlateau\n",
    "from tensorflow.keras import backend, models\n",
    "#from sklearn.model_selection import train_test_split  #could have used on the consolidated file.\n",
    "from sklearn.metrics import confusion_matrix\n",
    "\n",
    "from tensorflow.keras.applications import VGG16, MobileNet\n",
    "#from keras.applications.vgg16 import decode_predictions\n",
    "from keras.applications.vgg16 import preprocess_input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Declare the foundation of GA: Variable .\n",
    "population = [] # placeholders for individual/chromosomes.\n",
    "noPop = 6 # number of chromosomes in a population. Please note that this might\n",
    "# causing overhead to our computation\n",
    "fmap = [16,32,64,128,256] # each value is the features map value. this indicate\n",
    "# how much features map need to produce\n",
    "kernel = [2,3,5,7,9] # kernel is the window size to produce features map.\n",
    "activation = ['relu','sigmoid','tanh'] #this is the activation function to transform\n",
    "# features map before pooling\n",
    "dropout = [0.0,0.1,0.2,0.3,0.4,0.5] # this is dropout value\n",
    "learningRate = [0.0001,0.0005, 0.001, 0.005, 0.01,0.05,0.1,0.5] #this is\n",
    "# learning rate for cnn\n",
    "mutationRate = 0.3 #mutation rate for offspring to randomly change the allele value\n",
    "noGen = 10 # number of generation loop"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preparation and analysis of dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 35215 images belonging to 250 classes.\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# you store your datasets\n",
    "np.random.seed(12) # seeding\n",
    "\n",
    "train = ImageDataGenerator(rescale=1./255).flow_from_directory('dataset/train',target_size=(224,224))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Genetic Algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def initializingPopulation(population,noPop): # function to initializing chromosomes in population\n",
    "    for i in range (0,noPop): # ieach iteration will generate a chromosome\n",
    "        loss = 0.0\n",
    "        accuracy = 0.0\n",
    "        inPop = [choice(fmap),choice(kernel),choice(activation),choice(dropout),choice(learningRate),loss,accuracy]\n",
    "        # the command of choice is to randomly select the members of the list.\n",
    "        # loss and accuracy need to be set as 0.0 because the is no chromosome evaluation yet.\n",
    "        # in case of my project, loss and accuracy become my objective function/ fitness value\n",
    "        # the formula is min(Loss),max(Accuracy).\n",
    "        population.append (inPop)\n",
    "        # Now, append it to global list variable known as Population\n",
    "\n",
    "def rankSelection (population,noPop): # this is to ranking up and select half of the best chromosomes\n",
    "    population = sorted(population,key=itemgetter(5))\n",
    "    # code above is to ranking the lowest loss, which in ascending order\n",
    "    # in logical point of view, the lowest lost will be sorted at the top of population\n",
    "    population = sorted(population,key=itemgetter(6),reverse=True)\n",
    "    # code above is to ranking the highest accuracy, in descending order\n",
    "    # in logical point of view, the highest accuracy will be sorted at the top of population\n",
    "    noHalfPopulation = noPop//2 # this is to get half number of overall population\n",
    "    halfPopulation = population[:noHalfPopulation]\n",
    "    # in logic, the index of the list start with 0\n",
    "    # therefore, [:noHalfPopulation] denote select chromosomes higher than (<) noHalfPopulation\n",
    "    # example:\n",
    "    # total population of chromosomes is 20\n",
    "    # noHalfPopulation is 10\n",
    "    # [:noHalfPopulation] = [:10]\n",
    "    # this indicate that chromosomes with index 0-9 will be selected\n",
    "\n",
    "    return halfPopulation\n",
    "\n",
    "def tournamentSelection(selectedRank,tSize): # this is tournament selection function\n",
    "    #selectedRank is the placeholders of half best chromosomes\n",
    "    #tSize is tournament size\n",
    "    best = None # variable that hold the only best fitness value\n",
    "    for i in range(0,tSize):\n",
    "        # this loop is to make sure that the returned candidate isn't same for the second time\n",
    "        ind = choice(selectedRank) #randomly selected a candidate\n",
    "        if best == None:\n",
    "            # if variable beest does not hold any chromosome\n",
    "            best = ind\n",
    "            # the variable best will hold onto randomly selected chromosome\n",
    "        elif best == ind:\n",
    "            # if chromosome best and random are identical,\n",
    "            ind = choice(selectedRank)\n",
    "            # it then will randomly select another chromosome\n",
    "\n",
    "        if best[5]>ind[5] and best[6]<ind[6]:\n",
    "            # executed only if variable ind loss is lower and accuracy is higher than variable best\n",
    "            best = ind\n",
    "            #the chromosome in ind will replace chromosome in variable best\n",
    "\n",
    "    return best\n",
    "\n",
    "def crossover(p1,p2): # this is crossover function\n",
    "    crossoverPoint = len(p1)//2 # to get center of the chromosome\n",
    "    # however, this is my project.\n",
    "    # I advicing you guys to make randomly one point or k-point crossover\n",
    "    # here the logic that I can think of:\n",
    "    # crossoverPoint = random.random(0,4)\n",
    "    # the value in crossoverPoint then will become notion to slicing the chromosome\n",
    "    # as you guys can see there, the index is starting from 0 to 4\n",
    "    # I deliberately not include index 5 and 6\n",
    "    # this is because index of 5 is for loss\n",
    "    # index 6 is for accuracy\n",
    "    # both of it, in my project, is the fitness value\n",
    "    # it doesn't make any sense if those indexes become crossover cut point\n",
    "\n",
    "    c1 = p1[:crossoverPoint] + p2[crossoverPoint:]\n",
    "    # this is variable child 1\n",
    "    # p1[:crossoverPoint] indicates the head of parent 1\n",
    "    # p2[crossoverPoint:] indicate tail of parent 2\n",
    "    # :crossoverPoint means index lower than crossoverPoint will be selected as head\n",
    "    # crossoverPoint: indicate the index of crossoverPoint will be included as tail\n",
    "    # example:\n",
    "    # :6 means indexes of 0 to 5 will be selected as head\n",
    "    # 6: means indexes of 6 to 9 will be selected as tail\n",
    "\n",
    "    c2 = p2[:crossoverPoint] + p1[crossoverPoint:]\n",
    "    # this variable will become the second child\n",
    "    # parent 2 as head, parent 1 as tail\n",
    "\n",
    "    return c1, c2\n",
    "\n",
    "def mutation (c1,c2):\n",
    "    # this is mutation function\n",
    "    rate = round(random(),2)\n",
    "    print(c1)\n",
    "    print(c2)\n",
    "    \n",
    "    print ('rate: ' +str(rate))\n",
    "    # the rate variable will randomly generate value from 0 to 1\n",
    "    if rate<=mutationRate:\n",
    "        # this if-else statement is to execute mutation\n",
    "        # if variable rate lower or equal to variable mutationRate\n",
    "        # this if-else statement will execute\n",
    "        length = len(c1)-2\n",
    "        print ('length: '+str(length))\n",
    "        # the variable i will become limitation condition for randomly generated index\n",
    "        index = randint(0,length)\n",
    "        print('index: ' +str(index))\n",
    "        # this variable will generate random integer from 0 to lower than length of chromosome\n",
    "        # this exclude the index of loss and accuracy\n",
    "        # this is because mutating loss and accuracy doesn't make any sense\n",
    "\n",
    "        if index == 0:\n",
    "            c1[index] = choice(fmap)\n",
    "            c2[index] = choice(fmap)\n",
    "            # this is to change value of index 0 for both childs\n",
    "        elif index == 1:\n",
    "            c1[index] = choice(kernel)\n",
    "            c2[index] = choice(kernel)\n",
    "            # this is to change value of index 1 for both childs\n",
    "        elif index == 2:\n",
    "            c1[index] = choice(activation)\n",
    "            c2[index] = choice(activation)\n",
    "            # this is to change value of index 2 for both childs\n",
    "        elif index == 3:\n",
    "            c1[index] = choice(dropout)\n",
    "            c2[index] = choice(dropout)\n",
    "            # this is to change value of index 3 for both childs\n",
    "        elif index == 4:\n",
    "            c1[index] = choice(learningRate)\n",
    "            c2[index] = choice(learningRate)\n",
    "            # this is to change value of index 4 for both childs\n",
    "\n",
    "    return c1,c2\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generation 0\n",
      "[16, 9, 'tanh', 0.2, 0.1, 0.0, 0.0]\n",
      "[64, 7, 'sigmoid', 0.5, 0.05, 0.0, 0.0]\n",
      "[64, 2, 'sigmoid', 0.3, 0.05, 0.0, 0.0]\n",
      "[16, 7, 'sigmoid', 0.5, 0.5, 0.0, 0.0]\n",
      "[32, 2, 'relu', 0.0, 0.01, 0.0, 0.0]\n",
      "[64, 7, 'relu', 0.1, 0.0005, 0.0, 0.0]\n",
      "[16, 2, 'relu', 0.5, 0.005, 0.0, 0.0]\n",
      "[64, 5, 'tanh', 0.2, 0.1, 0.0, 0.0]\n",
      "[256, 9, 'tanh', 0.4, 0.005, 0.0, 0.0]\n",
      "[256, 2, 'tanh', 0.4, 0.0005, 0.0, 0.0]\n",
      "[16, 5, 'relu', 0.2, 0.001, 0.0, 0.0]\n",
      "[64, 3, 'tanh', 0.3, 0.0005, 0.0, 0.0]\n",
      "16 9 tanh 0.2 0.1\n",
      "Epoch 1/10\n",
      "100/100 - 37s - loss: 4941.0405 - accuracy: 0.0056\n",
      "Epoch 2/10\n"
     ]
    }
   ],
   "source": [
    "def cnn(population):\n",
    "    # this function is for cnn\n",
    "    # Making 2 different kind of file kinda headache for me\n",
    "    #but, if no one understand how it goes, later i will commit the separate files\n",
    "    for p in population:\n",
    "        #this is where the fitness value will evaluate\n",
    "        f = p[0] # features map variable\n",
    "        k = p[1] # kernel variable\n",
    "        a = p[2] # activation function variable\n",
    "        d = p[3] # dropout variable\n",
    "        l = p[4] # learning rate variable\n",
    "        print(f,k,a,d,l)\n",
    "        \n",
    "        #Let's try the mobilenet with ReduceLROnPlateau - 93% accuracy\n",
    "        backend.clear_session()\n",
    "\n",
    "        #Bring in the imagenet dataset training weights for the Mobilenet CNN model.\n",
    "        #Remove the classification top.\n",
    "        base_mobilenet = MobileNet(weights = 'imagenet', include_top = False, \n",
    "                           input_shape = (224,224,3))\n",
    "        base_mobilenet.trainable = False # Freeze the mobilenet weights.\n",
    "\n",
    "        model = Sequential()\n",
    "        model.add(base_mobilenet)\n",
    "        model.add(Conv2D(f,(k,k),input_shape=(224,224,3),padding='same'))\n",
    "        model.add(Activation(a))\n",
    "        model.add(MaxPooling2D(pool_size=(2,2),padding='same'))\n",
    "        model.add(Flatten())\n",
    "        model.add(Dense(32*32))\n",
    "        model.add(Dropout(d))\n",
    "        model.add(Dense(250))\n",
    "        model.add(Activation('softmax'))\n",
    "\n",
    "        model.compile(Adam(l),loss='categorical_crossentropy',metrics=['accuracy'])\n",
    "\n",
    "        history = model.fit(train,epochs=10,steps_per_epoch=100,verbose=2)\n",
    "        acc = np.mean(history.history['accuracy'])\n",
    "        loss = np.mean(history.history['loss'])\n",
    "        print(acc)\n",
    "        print(loss)\n",
    "\n",
    "        p[5] = loss\n",
    "        p[6] = acc\n",
    "        print(p[5])\n",
    "        print(p[6])\n",
    "\n",
    "# Here is how the code executed\n",
    "initializingPopulation(population,noPop)\n",
    "# Start with initializing population\n",
    "print ('Generation 0')\n",
    "for p in population:\n",
    "    print(p)\n",
    "\n",
    "for gen in range(0,noGen):\n",
    "    # looping for generation\n",
    "    # convolutional neural network fitness value\n",
    "    cnn(population)\n",
    "    # after cnn fitness evaluation, it will be ranked\n",
    "    selectedRank = rankSelection(population,noPop)\n",
    "    print('Selected Ranked Chromosomes:')\n",
    "    for sr in selectedRank:\n",
    "        print(sr)\n",
    "\n",
    "    if selectedRank[-1][5]<6 and selectedRank[-1][6]>0.6:\n",
    "        # this is  stopping condition for genetic algorithm.\n",
    "        # you can change or delete it if you want\n",
    "        # but, remember, waiting for 10 generation is time consumption\n",
    "        # and also computationally overhead\n",
    "        break\n",
    "    else:\n",
    "        child = [] #childs placeholders\n",
    "        tSize = 2 # this indicate 2-way tournament\n",
    "        numberOfMating = len(selectedRank)//2\n",
    "        # this is limiter  for number of child will generate\n",
    "        # let say, you have 12 total population\n",
    "        # let say, you have 6 ranking selection\n",
    "        # the mating process will be 3 time\n",
    "        # but, since the generated child is 2\n",
    "        # thus mating process would generate 6 children\n",
    "        # this is because number of mating * generated Children = half new population\n",
    "        # 3 mating process * 2 newly generated child = 6 population\n",
    "        for i in range (0,numberOfMating):\n",
    "            #this is where mating process take place\n",
    "            parent1 = tournamentSelection(selectedRank,tSize)\n",
    "            parent2 = tournamentSelection(selectedRank,tSize)\n",
    "            while parent2 is parent1:\n",
    "                parent2 = tournamentSelection(selectedRank,tSize)\n",
    "                #this is to make sure the candidate does not identical\n",
    "            c1,c2 = crossover(parent1,parent2)\n",
    "            print('Child 1: %s child 2: %s'%(c1,c2))\n",
    "            c1,c2 = mutation(c1,c2)\n",
    "            child.append(c1)\n",
    "            child.append(c2)\n",
    "\n",
    "        population.clear() # clearing old population\n",
    "        population = selectedRank + child # combining the best half of old population with new childs\n",
    "        # known as elitism method\n",
    "\n",
    "        print ('New Child' + str(gen))\n",
    "        for c in child:\n",
    "            print(c)\n",
    "\n",
    "        print('New Generation ' + str(gen+1))\n",
    "        for p in population:\n",
    "            print(p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
